controllers/storageclusterpeer/storageclusterpeer_controller.go

```diff
@@ -18,46 +18,488 @@ package storageclusterpeer
+				reflect.DeepEqual(oldObj.GetAnnotations(), newObj.GetAnnotations()))
+		},
+	}
+
+	secretPredicate := predicate.Funcs{
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return strings.HasPrefix(e.Object.GetName(), bootstrapSecretPrefix)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return strings.HasPrefix(e.ObjectOld.GetName(), bootstrapSecretPrefix)
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.StorageClusterPeer{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Owns(&rookCephv1.CephBlockPool{}, builder.WithPredicates(cephBlockPoolPredicate)).
+		Owns(&corev1.Secret{}, builder.WithPredicates(secretPredicate)).
+		Owns(&rookCephv1.CephRBDMirror{}).
+		Complete(r)
 }
 
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=ocs.openshift.io,resources=storageclusterpeers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephblockpools;,verbs=list;update;
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephrbdmirrors;,verbs=create;update
+//+kubebuilder:rbac:groups=core,resources=secrets,verbs=create;update;delete
+//+kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the StorageClusterPeer object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
-func (r *StorageClusterPeerReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+func (r *StorageClusterPeerReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	var err error
+	r.ctx = ctrllog.IntoContext(ctx, r.Log)
+	r.Log.Info("Reconciling StorageClusterPeer.")
+
+	// Fetch the StorageClusterPeer instance
+	r.storageClusterPeer = &v1.StorageClusterPeer{}
+	r.storageClusterPeer.Name = request.Name
+	r.storageClusterPeer.Namespace = request.Namespace
+
+	if err = r.get(r.storageClusterPeer); err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("StorageClusterPeer resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		r.Log.Error(err, "Failed to get StorageClusterPeer.")
+		return reconcile.Result{}, err
+	}
+
+	// Don't Reconcile the StorageClusterPeer if it is in failed state
+	if r.storageClusterPeer.Status.Phase == v1.StorageClusterPeerFailed {
```

Comment:
where are you setting the phase to be failed?

Reply:
removed this